spyrmsd/parallel.py

```diff
@@ -0,0 +1,118 @@
+try:
+    from pebble import ProcessPool
+except ImportError:
+    errmsg = (
+        "Parallel execution of SPyRMSD (`prmsdwrapper`) requires `pebble`."
+        "Please install `pebble`: https://github.com/noxdafox/pebble"
+    )
+    raise ImportError(errmsg)
+
+import os
+from concurrent.futures import TimeoutError
+from functools import partial
+from typing import List, Optional, Union
+
+import numpy as np
+
+from spyrmsd import molecule
+from spyrmsd.rmsd import rmsdwrapper
+
+
+def prmsdwrapper(
+    molrefs: Union[molecule.Molecule, List[molecule.Molecule]],
+    mols: Union[molecule.Molecule, List[molecule.Molecule]],
+    num_workers: Union[int, None] = 1,
+    symmetry: bool = True,
+    center: bool = False,
+    minimize: bool = False,
+    strip: bool = True,
+    cache: bool = True,
+    timeout: Optional[float] = None,
+    chunksize: int = 1,
+) -> List[float]:
+    """
+    Compute RMSD between two molecules with a timeout.
+    Parameters
+    ----------
+    molrefs: Union[molecule.Molecule, List[molecule.Molecule]]
+        Reference molecule
+    mols: Union[molecule.Molecule, List[molecule.Molecule]]
+        Molecules to compare to reference molecule
+    num_workers: int
+        Amount of processor to use for the parallel calculations
+    symmetry: bool, optional
+        Symmetry-corrected RMSD (using graph isomorphism)
+    center: bool, optional
+        Center molecules at origin
+    minimize: bool, optional
+        Minimised RMSD (using the quaternion polynomial method)
+    strip: bool, optional
+        Strip hydrogen atoms
+    timeout: float, optional
+        After how many seconds to stop the RMSD calculations
+    chunksize: int, optional
+        How many molecules to handle per process
+    Returns
+    -------
+    List[float]
+        RMSDs
+    """
+
+    # Ensure the num_workers is less or equal than the max number of CPUs.
+    # Silencing MyPy since os.cpu_count() can return None
+    num_workers = min(num_workers, os.cpu_count()) if os.cpu_count() is not None else 1  # type: ignore[type-var]
+
+    # Cast the molecules to lists if they aren't already
+    if not isinstance(molrefs, list):
```

Comment:
Parallelisation is performed over graph isomorphisms, so I don't think a single reference structure is particularly interesting to have. Is this for the timeout functionality only? If yes, I might be worth adding a note in the docstring, explaining the two (non-exclusive) functionalities of this function and that parallelisation only makes sense for multiple molecules.

Maybe it adds useless overhead, but it might be worth checking `num_workers` and raising a warning if a single molecules is passed with `num_workers > 1`? But maybe it's not worth it, because it would prevent to set `num_workers = None` by default.

Reply:
Yes, if there is only one structure it makes less sense to use this method, unless you want to use the timeout feature like you mentioned. Another scenario could be where you want to check just one structure against a lot of reference structures, which would also be executed in parallel. (For example if you docked a structure 100 times, and want to compare the top structure directly with all the other compounds, it would treat every RMSD calculation seperately and run in parallel across multiple cores). In this case, it would basically match the single input structure to all the reference molecules for you, and calculate it seperately using `rmsdwrapper`. So `rmsdwrapper(mol, refmol1)`, `rmsdwrapper(mol, refmol2)`, ... is computed across all the available cores.

Reply:
> Another scenario could be where you want to check just one structure against a lot of reference structures, which would also be executed in parallel.

Yes, but in this case you would have to re-compute the graph isomorphism each time, which might be the most expensive thing. I'm starting to wonder if these two cases (one reference and many many corresponding structures, or multiple references with few corresponding structures) are different enough to distinguish them? The first case would definitely benefit to compute the graphs isomorphisms once, and then parallelise over the RMSD calculations (although these are quite fast, so the overhead of parallelisation might be too much...).

Reply:
What do you think? Do you have a feel about how much is the parallel overhead compared to just the RMSD calculation?

Reply:
Yeah so the parallel method is "stupid" in the sense that it re-computes the graph isomorphism each time (which I agree is not efficient). However I think that for this there would need to be made some more changes (maybe cache it somehow so it can be reused?). Also the graph isomorphism calculations aren't perfomed in a multicore manner either. In an ideal world the isormorphisms would be calculated using all the assigned num_workers, and be cached. After that the RMSDs could be computed across all num_workers (single core) and reuse the cached isomorphisms until all calculations are finished. However I don't have the bandwidth to work on that. The initial goal for me was to have a function with `timeout` function so it can run more stable on libraries that have "problematic" molecules (like Drugbank).
Also for your question about overhead: with pebble there still is some overhead, but compared to spawning the processes manually as I did in the first try, it looks to be relatively small (but ofcourse not 0). I would say around a 10-20% overhead maybe? Here are the benchmarks I shared before:

- regular `rmsdwrapper` function: 12.00s
- `pebblermsdwrapper` with num_worker=1: 14.27s
- `pebblermsdwrapper` with num_worker=2: 7.69s
- `pebblermsdwrapper` with num_worker=4: 4.93s

So using 2 Cores already gives you a faster performance than 1 (which in my opinion seems already worth it, especially with the timeout feature)

Reply:
> However I think that for this there would need to be made some more changes (maybe cache it somehow so it can be reused?).

Yes, this is how a standard RMSD calculation already works, if you are passing a single reference and many structures to compare it with. The isomorphism is computed only once and cached:

https://github.com/RMeli/spyrmsd/blob/62c5df31a5926cebfc7ae5bb4ab3a661e37443f2/spyrmsd/rmsd.py#L239-L240

So one would have to use `_rmsd_isomorphic_core` in a similar way in order to implement parallelisation over RMSD calculations with a single graph isomorphism.

> However I don't have the bandwidth to work on that.

Totally understandable. And so that there are no misunderstanding, I'm not suggesting to implement both directly, but to narrow the scope of the one you are implementing (parallelisation on graph isomorphism, and timeout), and leave the other less interesting case (parallelisation over RMSD calculations with the same isomorphism) as a TODO.

> Also for your question about overhead: with pebble there still is some overhead, but compared to spawning the processes manually as I did in the first try, it looks to be relatively small (but ofcourse not 0).

Yes, I was wondering how the overhead compares with RMSD calculations only, instead of graph isomorphism + RMSD. The former is vectorised with `numpy`, so it should be very fast (and the overhead would have much more impact).